jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/handler/EventsHandler.java

```diff
@@ -0,0 +1,419 @@
+//
+
+package org.eclipse.jetty.server.handler;
+
+import java.nio.ByteBuffer;
+import java.util.concurrent.TimeoutException;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+import org.eclipse.jetty.http.HttpFields;
+import org.eclipse.jetty.http.MetaData;
+import org.eclipse.jetty.io.Content;
+import org.eclipse.jetty.server.Handler;
+import org.eclipse.jetty.server.HttpStream;
+import org.eclipse.jetty.server.Request;
+import org.eclipse.jetty.server.Response;
+import org.eclipse.jetty.util.Callback;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * <p>A {@link Handler.Wrapper} that fires events during the processing of the requests.</p>
+ * <p>EventsHandler will emit events for the various phases the server goes through while
+ * processing an HTTP request and response.</p>
+ * <p>Subclasses may listen to those events to track timing and/or other values such as
+ * request URI, etc.</p>
+ * <p>The events parameters, especially the {@link Request} object, may be
+ * in a transient state depending on the event, and not all properties/features
+ * of the parameters may be available inside a listener method.</p>
+ * <p>It is recommended that the event parameters are <em>not</em> acted upon
+ * in the listener methods, or undefined behavior may result. On the other
+ * hand, it is legit to store request attributes in one listener method that
+ * may be possibly retrieved in another listener method in a later event.</p>
+ * <p>Listener methods are invoked synchronously from the thread that is
+ * performing the request processing, and they should not call blocking code
+ * (otherwise the request processing will be blocked as well).</p>
+ * <p>The kind of chunk passed to {@link #onRequestRead(Request, Content.Chunk)} depends on
+ * the parent of this handler. For instance, if the parent is the Server, then raw chunks
+ * are always passed. If somewhere in the parent chain is the {@code GzipHandler} then
+ * unzipped chunks are passed.</p>
+ */
+public abstract class EventsHandler extends Handler.Wrapper
+{
+    private static final Logger LOG = LoggerFactory.getLogger(EventsHandler.class);
+    
+    public EventsHandler()
+    {
+    }
+
+    public EventsHandler(Handler handler)
+    {
+        super(handler);
+    }
+
+    @Override
+    public boolean handle(Request request, Response response, Callback callback) throws Exception
+    {
+        // Before handling
+        ReadOnlyRequest roRequest = new ReadOnlyRequest(request);
+        fireOnBeforeHandling(roRequest);
+        try
+        {
+            EventsRequest wrappedRequest = new EventsRequest(request, roRequest);
+            EventsResponse wrappedResponse = new EventsResponse(roRequest, response);
+            request.addHttpStreamWrapper(stream -> new HttpStream.Wrapper(stream)
+            {
+                private void fireNotifications(Throwable failure)
```

Comment:
"fire" and "notify" are the same thing.  See my other comment about the naming... this could just be "notify".    I'll say no more :)